include/boost/beast/core/impl/buffers_adaptor.hpp

```diff
@@ -34,332 +35,162 @@ namespace beast {
+        , first_offset_(other.first_offset_)
+        , last_size_(other.last_size_)
     {
     }
 
     readable_bytes& operator=(
         readable_bytes const& other)
     {
-        b_ = other.b_;
+        first_ = other.first_;
+        last_ = other.last_;
+        first_offset_ = other.first_offset_;
+        last_size_ = other.last_size_;
         return *this;
     }
 #else
-    readable_bytes(
-        readable_bytes const&) = default;
-    readable_bytes& operator=(
-        readable_bytes const&) = default;
+    subrange(
+        subrange const&) = default;
+    subrange& operator=(
+        subrange const&) = default;
 #endif
 
-    template<bool isMutable_ = isMutable, class =
-        typename std::enable_if<! isMutable_>::type>
-    readable_bytes(
-        readable_bytes<true> const& other) noexcept
-        : b_(other.b_)
+    // allow conversion from mutable to const
+    template<bool thisIsMutable = isMutable,
+        typename std::enable_if<!thisIsMutable>::type * = nullptr>
+    subrange(subrange<true> const &other)
+        : first_(other.first())
+        , last_(other.last())
+        , first_offset_(other.first_offset())
+        , last_size_(other.last_size())
     {
     }
 
-    template<bool isMutable_ = isMutable, class =
-        typename std::enable_if<! isMutable_>::type>
-    readable_bytes& operator=(
-        readable_bytes<true> const& other) noexcept
-    {
-        b_ = other.b_;
-        return *this;
-    }
+    struct iterator;
 
-    const_iterator
+    iterator
     begin() const;
 
-    const_iterator
+    iterator
     end() const;
 
-private:
-    friend class buffers_adaptor;
-
-    readable_bytes(buffers_adaptor const& b)
-        : b_(&b)
-    {
-    }
-};
-
-#if BOOST_WORKAROUND(BOOST_MSVC, < 1910)
-# pragma warning (pop)
-#endif
-
-//------------------------------------------------------------------------------
-
-template<class MutableBufferSequence>
-template<bool isMutable>
-class buffers_adaptor<MutableBufferSequence>::
-      readable_bytes<isMutable>::
-      const_iterator
-{
-    iter_type it_{};
-    buffers_adaptor const* b_ = nullptr;
-
-public:
-    using value_type = typename
-        std::conditional<isMutable,
-            net::mutable_buffer,
-            net::const_buffer>::type;
-    using pointer = value_type const*;
-    using reference = value_type;
-    using difference_type = std::ptrdiff_t;
-    using iterator_category =
-        std::bidirectional_iterator_tag;
-
-    const_iterator() = default;
-    const_iterator(const_iterator const& other) = default;
-    const_iterator& operator=(const_iterator const& other) = default;
+    value_type
+    get_buffer(iter_type it) const;
 
-    bool
-    operator==(const_iterator const& other) const
-    {
-        return b_ == other.b_ && it_ == other.it_;
-    }
-
-    bool
-    operator!=(const_iterator const& other) const
-    {
-        return !(*this == other);
-    }
+    iter_type
+    first() const;
```

Comment:
`get_buffer`, `first`, `last`, `first_offset`, and `last_size` should not be public, since they are not public interfaces.